### PR TITLE
Fix CPU broadcast exchanges being left unreplaced due to AQE and reuse [databricks]

### LINKED ISSUE #7818: [BUG] Reuse of broadcast exchange can lead to unnecessary CPU fallback
On some NDS queries in Databricks the plan is falling back to the CPU for a reused broadcast exchange, but oddly the reuse is sometimes not actually a reuse at all.  Take the following plan UI snippet as an example:

image

Note on the right side how it has a ReusedExchange that does not point to a shuffle that is reused.

The broadcast that is on the CPU does not have anything about it that would normally cause a CPU fallback.  The fallback occurs because of the incoming CPU exchange which itself should have been replaced.  What appears to be happening is that the RAPIDS Accelerator is replacing a CPU exchange with a GPU exchange and that CPU exchange was being reused elsewhere in the plan (not visible at the time of RAPIDS Accelerator replacement due to AQE).

To repro, run NDS query 40 on Databricks 10.4

### PR DESCRIPTION
**Pull request by jlowe**
Fixes #7818.

On some Spark platforms, the CPU planning is caching CPU broadcast exchanges and reusing them even after they were replaced by a GPU broadcast exchange.  There's also cases during AQE planning where GPU broadcasts are not being reused after being planned, so this adds another pass to look for opportunities to reuse when possible.  Unfortunately AQE planning prevents us from replacing a GPU broadcast exchange at the top of the plan with a ReusedExchangeExec, since there's an assert in AdaptiveSparkPlan exec that an exchange remains an exchange (and a ReusedExchangeExec is not an exchange, sadly).

### CODE REVIEW

**Comment by jlowe** [comment]
Manually verified this fixed the CPU broadcast fallbacks on Databricks 10.4 with various NDS queries.  I was unable to find a small-scale test that could replicate the issue, so I added a unit test instead.

**Comment by jlowe** [comment]
build

**Comment by jlowe** [comment]
build

**Comment by pxLi** [comment]
build

**Comment by pxLi** [comment]
build

**Comment by pxLi** [comment]
retriggered build, previous timeout due to internal networking slowness

**Comment by pxLi** [comment]
build

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuBroadcastExchangeExec.scala [hunk 1]
```diff
@@ -537,9 +539,45 @@ object GpuBroadcastExchangeExecBase {
...
+    (val cpuCanonical: BroadcastExchangeExec)
     extends GpuBroadcastExchangeExecBase(mode, child) {
+
+  override def otherCopyArgs: Seq[AnyRef] = Seq(cpuCanonical)
+
+  private var _isGpuPlanningComplete = false
+
+  def isGpuPlanningComplete: Boolean = _isGpuPlanningComplete
+
+  def markGpuPlanningComplete(): Unit = {
```
Comment (revans2) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuBroadcastExchangeExec.scala:554]: Would some docs here help? This part of the code is not totally clear as to why this is being used/done.

**Review by revans2 — APPROVED** [review]
Really just a nit for some scala docs. Also if this is only an issue for databricks do we want to have it enabled by that specific shim?

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuBroadcastExchangeExec.scala (hunk 1 above)
Comment (jlowe) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuBroadcastExchangeExec.scala:563]: Updated to add some comments here.

**Comment by jlowe** [comment]
> if this is only an issue for databricks do we want to have it enabled by that specific shim?

I thought about this, but I left it as a common optimization for two reasons:
1. We'll get more testing overall if it runs on all platforms
1. If this optimization does become applicable to other platforms, it seems like that would be a net benefit.

But I'm torn on it as well.  Happy to switch it to something specific to Databricks (or even specific to spark321db) if desired.

**Comment by jlowe** [comment]
build

**Pull request merged**